src/context/TodosContext.tsx

```diff
@@ -0,0 +1,276 @@
+        todos: state.todos.filter(todo => todo.id !== action.payload),
+      };
+
+    case 'loading':
+      return { ...state, isLoading: action.payload };
+
+    case 'isEditingProcessing':
+      return { ...state, isEditingProcessing: action.payload };
+
+    case 'isDeletingTodo':
+      return { ...state, isDeletingTodo: action.payload };
+
+    case 'todos/loaded':
+      return { ...state, isLoading: false, todos: action.payload };
+
+    case 'todos/setFilterStatus':
+      return {
+        ...state,
+        filterStatus: action.payload,
+      };
+
+    case 'todos/setCurrentId':
+      return {
+        ...state,
+        currentId: action.payload ? [...state.currentId, action.payload] : [],
+      };
+
+    case 'todos/setTempTodo':
+      return {
+        ...state,
+        tempTodo: action.payload,
+      };
+
+    case 'todos/setError':
+      return { ...state, todosError: action.payload };
+
+    case 'rejected':
+      return { ...state, isLoading: false, todosError: action.payload };
+    default:
+      return state;
+  }
+}
+
+const TodosProvider: React.FC<Props> = ({ children }) => {
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const {
+    todos,
+    filterStatus,
+    isLoading,
+    todosError,
+    tempTodo,
+    currentId,
+    isEditingProcessing,
+    isDeletingTodo,
+  } = state;
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      dispatch({ type: 'loading', payload: true });
+      try {
+        const fetchedTodos = await getTodos();
+
+        dispatch({ type: 'todos/loaded', payload: fetchedTodos });
+      } catch {
+        dispatch({
+          type: 'rejected',
+          payload: 'Unable to load todos',
+        });
+
+        wait(3000).then(() =>
+          dispatch({ type: 'todos/setError', payload: '' }),
+        );
+      } finally {
+        dispatch({ type: 'loading', payload: false });
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  const addTodo = (todo: Todo) => {
+    dispatch({ type: 'todos/addTodo', payload: todo });
+  };
+
+  const handleUpdateTodo = (id: TodoId, todo: Todo) => {
+    dispatch({ type: 'todos/updateTodo', payload: { id, todo } });
+  };
+
+  const handleDeleteTodo = (id: TodoId) => {
+    dispatch({ type: 'todos/delete', payload: id });
+  };
+
+  const setTempTodo = (todo: Todo | null) => {
+    dispatch({ type: 'todos/setTempTodo', payload: todo });
+  };
+
+  const handleFilterTodo = (status: string) => {
+    dispatch({ type: 'todos/setFilterStatus', payload: status });
+  };
+
+  const handleSetError = (errorMessage: string) => {
+    dispatch({ type: 'todos/setError', payload: errorMessage });
+
+    wait(3000).then(() => dispatch({ type: 'todos/setError', payload: '' }));
+  };
+
+  const handleToggleTodoCheck = async (
+    todoId: TodoId,
+    todo: Todo,
+    toggleVal: boolean,
+  ) => {
+    dispatch({ type: 'loading', payload: true });
+    dispatch({ type: 'todos/setCurrentId', payload: todoId });
```

Comment:
You can use only one state with todo ids that currently loading, it will be much easier to keep in mind and maintain